noir-projects/aztec-nr/aztec/src/context/private_context.nr

```diff
@@ -208,24 +208,19 @@ impl PrivateContext {
     }
 
     pub fn request_nsk_app(&mut self, npk_m_hash: Field) -> Field {
-        // A value of empty nullifier keys will fail the key validation request.
-        let cached_nullifier_keys = self.nullifier_key.unwrap_or(NullifierKeys::empty());
+        let cached_request = self.last_nullifier_key_validation_request.unwrap_or(NullifierKeyValidationRequest::empty());
 
-        let nullifier_keys = if cached_nullifier_keys.master_nullifier_public_key.hash() == npk_m_hash {
-            cached_nullifier_keys
+        if cached_request.master_nullifier_public_key.hash() == npk_m_hash {
+            // We get a match so the cached request is the latest one 
+            cached_request.app_nullifier_secret_key
         } else {
-            let fetched_nullifier_keys = get_nullifier_keys(npk_m_hash);
-            let request = NullifierKeyValidationRequest {
-                master_nullifier_public_key: fetched_nullifier_keys.master_nullifier_public_key,
-                app_nullifier_secret_key: fetched_nullifier_keys.app_nullifier_secret_key
-            };
+            // We didn't get a match meaning the cached result is stale. We fetch new values from oracle and instruct
+            // protocol circuits to validate them by storing the validation request in context.
+            let request = get_nullifier_key_validation_request(npk_m_hash);
             self.nullifier_key_validation_requests.push(request);
-            self.nullifier_key = Option::some(fetched_nullifier_keys);
-            fetched_nullifier_keys
-        };
-
-        assert_eq(nullifier_keys.master_nullifier_public_key.hash(), npk_m_hash);
```

Comment:
I think you still need this assertion. Without it, the request might be valid but not matching `npk_m_hash`. Beneficial to add a comment to make it clearer. We can likely move the assertion up. When we are looking up the cache we have already checked that `npk_m_hash` matches.

Reply:
But why would we need it when it's used only in unconstrained context? And it constrains only half the response (not the nsk_app) so I felt like it mostly just gives the illusion of it being constrained.

Reply:
Fixed in https://github.com/AztecProtocol/aztec-packages/commit/776538f59928a809aab4b0bd62300c175e11cd17